### PR TITLE
Recording options and Bug Fixes

### LINKED ISSUE #32: AudioManager.Current.CreatePlayer only works with stream, not filename
I first tried to use audioPath directly but it would not work.  It does work with the stream.  So I am good but the overloaded version that takes filename seems to have a problem.

         IAudioPlayer_audioPlayer = AudioManager.Current.CreatePlayer(audioPath); // does not work
        _audioPlayer.Play();

         var stream = File.OpenRead(audioPath);
        IAudioPlayer_audioPlayer = AudioManager.Current.CreatePlayer(stream); // works
        _audioPlayer.Play();

### LINKED ISSUE #49: Allow customizing the sample rate
Also realised can't set the sample rate

_Originally posted by @SineVector241 in https://github.com/jfversluis/Plugin.Maui.Audio/issues/41#issuecomment-1366319735_

### LINKED ISSUE #73: Speed issues on iOS and Windows
When I set   _audioPlayer.Speed = 0.6 it works on android as expected. On iOS  _audioPlayer.Speed is set to 0.60000123 but it sounds at normal speed.  On Windows usually it takes from 2 to 10 attempts to set _audioPlayer.Speed  to desired value. For widows I use code like this: 

```
 for (int attempt = 1; attempt <= 20; attempt++)
 {
     _audioPlayer.Speed = 0.6;
    
     if (Math.Round(_audioPlayer.Speed, 1) == 0.6)
     {
         break;
     }
     else
     {
         await Task.Delay(10);
     }
 }
```

### LINKED ISSUE #103: Temp audio file never deleted on Android
The raw audio is written to a temp file on Android. This file is not deleted.

`
static string GetTempFilePath()
{
	return Path.Combine("/sdcard/", Path.GetTempFileName());
}
` 
The path to the Audio file should probably also be replaced, maybe with `FileSystem.CacheDirectory`?

### LINKED ISSUE #104: Audio Player is not playing from stream on iOS after recording, and player.IsPlaying is always false
Hi team,
I am using this plugin for recording and playing sound.
When i click play button to play sound that time i am adding 
[WP_Recorder (2).zip](https://github.com/jfversluis/Plugin.Maui.Audio/files/14649372/WP_Recorder.2.zip)

You can check RecordAudio.xaml view which is to Record and Play Audio. 

Here when Play duration ends after that PlaybackEnded event is not working and is not called. Everything works very well for android.
iOS having issue to play sound etc.

### PR DESCRIPTION
**Pull request by borrmann**
Adds support for RecordingOptions and addresses the following issues

Fixes #103 
Fixes #104 
Fixes #73 (for iOS)
Fixes #49
Fixes #32 (for Android, iOS seemed to have worked already, did not look into windows)

I have added the fix to set the speed on iOS, but I think an API change could be considered for setting the speed in the future. 
On android setting the speed is via a method and not a setter, so I think it should be changed to a method instead a setter as well.

### CODE REVIEW

**Comment by jfversluis** [comment]
Wow great work @borrmann I will review asap!

File: samples/Plugin.Maui.Audio.Sample/Plugin.Maui.Audio.Sample.csproj [hunk 1]
```diff
@@ -30,6 +30,11 @@
 		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'tizen'">6.5</SupportedOSPlatformVersion>
 	</PropertyGroup>
 
+	<PropertyGroup Condition="'$(TargetFramework)'=='net8.0-ios'">
+	  <CodesignKey>Apple Development: Max Borrmann (8262ZJ26CU)</CodesignKey>
+	  <CodesignProvision>VS: WildCard Development</CodesignProvision>
+	</PropertyGroup>
+
```
Comment (jfversluis) [inline samples/Plugin.Maui.Audio.Sample/Plugin.Maui.Audio.Sample.csproj:37]:
```suggestion
```

File: samples/Plugin.Maui.Audio.Sample/ViewModels/AudioRecorderPageViewModel.cs [hunk 2]
```diff
@@ -1,8 +1,11 @@
-﻿using System.Diagnostics;
+﻿using System.Collections.ObjectModel;
+using System.Diagnostics;
 using static Microsoft.Maui.ApplicationModel.Permissions;
 
 namespace Plugin.Maui.Audio.Sample.ViewModels;
 
+
+
```
Comment (jfversluis) [inline samples/Plugin.Maui.Audio.Sample/ViewModels/AudioRecorderPageViewModel.cs:8]:
```suggestion
```

File: src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs [hunk 3]
```diff
@@ -81,6 +85,31 @@ internal AudioPlayer(string fileName)
...
+	static void InitAudioSession()
+	{
+		var audioSession = AVAudioSession.SharedInstance();
+
+		// AVAudioSessionCategoryOptions.DefaultToSpeaker ensures no other audio is playing
+		var error = audioSession.SetCategory(AVAudioSessionCategory.Playback, AVAudioSessionCategoryOptions.DefaultToSpeaker);
+		if (error is not null)
+		{
+			Trace.TraceWarning(error.ToString());
+			//throw new FailedToPlayAudioException(error.ToString());
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs:97]:
```suggestion
```

File: src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs [hunk 4]
```diff
@@ -81,6 +85,31 @@ internal AudioPlayer(string fileName)
...
+		}
+
+		error = audioSession.SetActive(true, AVAudioSessionSetActiveOptions.NotifyOthersOnDeactivation);
+		if (error is not null)
+		{
+			Trace.TraceWarning(error.ToString());
+			//throw new FailedToPlayAudioException(error.ToString());
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs:104]:
```suggestion
```

File: src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.android.cs [hunk 5]
```diff
@@ -157,7 +196,10 @@ void CopyWaveFile(string? sourcePath, string destinationPath)
 				outputStream.Close();
 			}
 		}
-		catch { }
+		catch 
+		{ 
+			// this should maybe throw an exception?
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.android.cs:201]: At the very least lets add something to a logger

File: src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.macios.cs [hunk 6]
```diff
@@ -12,36 +14,41 @@ partial class AudioRecorder : IAudioRecorder
...
+			Trace.TraceWarning(error.ToString());
+			//throw new FailedToStartRecordingException(error.ToString());
 		}
 
-		error = audioSession.SetActive(true);
+		AVAudioSession.SharedInstance().SetActive(true, AVAudioSessionSetActiveOptions.NotifyOthersOnDeactivation, out error);
 		if (error is not null)
 		{
-			throw new Exception(error.ToString());
+			Console.WriteLine(error.ToString());
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.macios.cs:31]: Can we put this in a logger? Additionally to the console?

File: src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs [hunk 7]
```diff
@@ -26,6 +26,20 @@ public interface IAudioRecorder
 	///<param name="filePath">The path on disk to store the recording.</param>
 	Task StartAsync(string filePath);
 
+	///<Summary>
+	/// Start recording audio to disk in a randomly generated file. AudioRecordingOptions are only supported on Android and iOS.
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs:30]: Just out of curiosity, is it not possible on Windows or we just don't know how?

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	public const int DefaultSampleRate = 44100;
+	public const ChannelType DefaultChannels = ChannelType.Mono;
+	public const BitDepth DefaultBitDepth = BitDepth.Pcm16bit;
+	public const Encoding DefaultEncoding = Encoding.LinearPCM;
+
+	
+
+	/// <summary>
+	/// Sample rate of the audio recording
+	/// 44100Hz is currently the only rate that is guaranteed to work on all devices on android, <see href="https://developer.android.com/reference/android/media/AudioRecord.html">android audio record documentation</see>
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:26]:
```suggestion
	/// Sample rate of the audio recording.
	/// 44,100Hz is currently the only rate that is guaranteed to work on all devices on Android, <see href="https://developer.android.com/reference/android/media/AudioRecord.html">Android audio record documentation</see>
```

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public int SampleRate { get; set; } = DefaultSampleRate;
+
+	/// <summary>
+	/// Channels of the audio recording.
+	/// MONO is guaranteed to work on all devices on android, <see href="https://developer.android.com/reference/android/media/AudioRecord.html">android audio record documentation</see>
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:32]:
```suggestion
	/// Mono is guaranteed to work on all devices on Android, <see href="https://developer.android.com/reference/android/media/AudioRecord.html">Android audio record documentation</see>
```

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public ChannelType Channels { get; set; } = DefaultChannels;
+
+	/// <summary>
+	/// Bit depth of the audio recording
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:37]:
```suggestion
	/// Bit depth of the audio recording.
```

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public BitDepth BitDepth { get; set; } = DefaultBitDepth;
+
+	/// <summary>
+	/// encoding type of the audio recording.
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:42]:
```suggestion
	/// Encoding type of the audio recording.
```

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public Encoding Encoding { get; set; } = DefaultEncoding;
+
+	/// <summary>
+	/// If true, will throw an exception if the recording options are not supported
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:47]:
```suggestion
	/// Gets or sets whether the functionality will thrown an exception if the configured recording options are not supported.
```

**Review by jfversluis — COMMENTED** [review]
Some minor things for the first pass!

**Comment by bijington** [comment]
This is going to conflict with https://github.com/jfversluis/Plugin.Maui.Audio/pull/101 in some areas. I would personally prefer to get 101 in first and then that solves the recording bug. I'll be happy to assist with whatever merge conflicts arise from that.

File: src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs (hunk 7 above)
Comment (borrmann) [reply src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs:30]: I just don't know. But I can look into it!

**Comment by jfversluis** [comment]
OK so some conflicts here because I merged the other one, but I'd really love to have this!

Can we work these out? @borrmann are you still interested? Else we'll just take it from here and get this over the finish line :)

**Comment by bijington** [comment]
> OK so some conflicts here because I merged the other one, but I'd really love to have this!
> 
> 
> 
> Can we work these out? @borrmann are you still interested? Else we'll just take it from here and get this over the finish line :)

Considering my PR caused the conflicts in more than happy to help out if required

**Comment by borrmann** [comment]
@jfversluis @bijington no worries. I'll look into it asap. Unless you prefer do take control, no issue with that either.

**Comment by jfversluis** [comment]
I'd love for you to take it home as its your brain child! But if you feel it's going to take too long (whatever that is) or you don't have the time right now please just let us know and we'll take it!

**Comment by borrmann** [comment]
@bijington I merged your changes into my PR. When starting a Recording session, the app crashes on iOS with:

```
[0:] An error occurred: 'The operation couldn’t be completed. (OSStatus error -50.)'. Callstack: '   at Plugin.Maui.Audio.ActiveSessionHelper.InitializeSession(BaseOptions options) in C:\Users\maxbo\source\repos\borrmann\Plugin.Maui.Audio\src\Plugin.Maui.Audio\ActiveSessionHelper.macios.cs:line 15
   at Plugin.Maui.Audio.AudioRecorder.StartAsync(String filePath, AudioRecordingOptions options) in C:\Users\maxbo\source\repos\borrmann\Plugin.Maui.Audio\src\Plugin.Maui.Audio\AudioRecorder\AudioRecorder.macios.cs:line 43
   at Plugin.Maui.Audio.AudioRecorder.StartAsync() in C:\Users\maxbo\source\repos\borrmann\Plugin.Maui.Audio\src\Plugin.Maui.Audio\AudioRecorder\AudioRecorder.macios.cs:line 33
   at Plugin.Maui.Audio.Sample.ViewModels.AudioRecorderPageViewModel.Start() in C:\Users\maxbo\source\repos\borrmann\Plugin.Maui.Audio\samples\Plugin.Maui.Audio.Sample\ViewModels\AudioRecorderPageViewModel.cs:line 170
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at Foundation.NSAsyncSynchronizationContextDispatcher.Apply() in /Users/builder/azdo/_work/1/s/xamarin-macios/src/Foundation/NSAction.cs:line 176
   at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:line 58
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:line 94
   at Plugin.Maui.Audio.Sample.Program.Main(String[] args) in C:\Users\maxbo\source\repos\borrmann\Plugin.Maui.Audio\samples\Plugin.Maui.Audio.Sample\Platforms\iOS\Program.cs:line 13
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)'
```

I don't think line 15 in ActiveSessionHelper should throw an error, and rather trace a warning

**Comment by bijington** [comment]
@borrmann Sorry I hope I haven't left things in a state for you. Let me know if I can assist with any merging, etc.

**Comment by borrmann** [comment]
no problem, I replaced it with Trace (for now) to ensure there is no app crash so I could test my changes. I think it makes sense you take it from there, since the error could mean, that the options are not always applied as intended, but that should not effect this PR. I saw you were planning to add some features to the sample to play around with the different options anyways, so that should help to figure out if the error does effect the behavior.

Aside from that, I did quick tests on Windows, iOS and Android (Emulator). I think the changes from this PR work as inteded.

@bijington @jfversluis let me know if anything else is needed for this. I did notice Windows does not reliably play the audio file. If it does, position and duration are not updated. I don't think that has something to do with my changes here, since this shoul not have effected the Windows Audio player at all. If windows works as expected on your main, please let me know. 

However, I think I might have worked on these issues in my other PR, so I will update that and check whether the changes there resolved these issues already.

**Comment by bijington** [comment]
Thanks. I'll have to investigate the issue as I suspect something must be wrong there. I can do that outside of this PR as you suggested

**Comment by bijington** [comment]
> Thanks. I'll have to investigate the issue as I suspect something must be wrong there. I can do that outside of this PR as you suggested

I have fixed the issue here: https://github.com/jfversluis/Plugin.Maui.Audio/pull/111

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	/// </summary>
+	public Encoding Encoding { get; set; } = DefaultEncoding;
+
+	/// <summary>
+	/// Gets or sets whether the functionality will thrown an exception if the configured recording options are not supported.
+	/// </summary>
+	public bool ThrowIfNotSupported { get; set; } = false;
+
+
+
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:52]:
```suggestion
	public bool ThrowIfNotSupported { get; set; } = false;
```

File: src/Plugin.Maui.Audio/AudioRecordingOptions.cs [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+public class AudioRecordingOptions
+{
+	public const int DefaultSampleRate = 44100;
+	public const ChannelType DefaultChannels = ChannelType.Mono;
+	public const BitDepth DefaultBitDepth = BitDepth.Pcm16bit;
+	public const Encoding DefaultEncoding = Encoding.LinearPCM;
+
+	
+
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecordingOptions.cs:23]:
```suggestion
	public const Encoding DefaultEncoding = Encoding.LinearPCM;
```

File: src/Plugin.Maui.Audio/AudioRecorder/Primitives/Encoding.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Plugin.Maui.Audio;
+
+
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/Primitives/Encoding.cs:3]:
```suggestion
﻿namespace Plugin.Maui.Audio;
```

File: src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs [hunk 16]
```diff
@@ -26,6 +26,20 @@ public interface IAudioRecorder
...
+	public Task StartAsync(AudioRecordingOptions options) => Task.CompletedTask;
+
+	///<Summary>
+	/// Start recording audio to disk in the supplied <paramref name="filePath"/>.
+	///</Summary>
+	///<param name="filePath">The path on disk to store the recording. AudioRecordingOptions are only supported on Android and iOS. If options are used, read the audio stream and write your own file. The default header might not match your options.</param>
+	///<param name="options">The audio recording options.</param>
+	public Task StartAsync(string filePath, AudioRecordingOptions options) => Task.CompletedTask;
+
+
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs:42]:
```suggestion
	public Task StartAsync(string filePath, AudioRecordingOptions options) => Task.CompletedTask;

```

File: src/Plugin.Maui.Audio/AudioRecorder/Primitives/Encoding.cs (hunk 15 above)
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/Primitives/Encoding.cs:3]:
```suggestion
﻿namespace Plugin.Maui.Audio;
```

**Review by jfversluis — APPROVED** [review]

**Comment by jfversluis** [comment]
Thank you so much for this and seeing this through @borrmann!

**Pull request merged**